src/main/trial.ts

```diff
@@ -0,0 +1,115 @@
+import { Logger } from '@nodescript/logger';
+import { Redis } from 'ioredis';
+import { config } from 'mesh-config';
+import { dep } from 'mesh-ioc';
+
+import { AccessForbidden } from './ac-auth.js';
+
+export interface TrialServiceRestriction {
+    serviceName: string;
+    requestCount: number;
+}
+
+export interface Trial {
+    serviceRestrictions: Array<TrialServiceRestriction>;
+}
+
+export interface TokenServiceRestriction {
+    serviceName: string;
+    requestLimit: number;
+}
+
+export class TrialClient {
+
+    @config() private REDIS_URL!: string;
+    @config({ default: 'cache:framework:trialClient' }) private TRIAL_KEY_PREFIX!: string;
+    @dep() private logger!: Logger;
+
+    private isRunning = false;
+    redisClient: Redis;
+
+    constructor() {
+        this.redisClient = this.createRedisClient();
+    }
+
+    async start() {
+        if (this.isRunning) {
+            return;
+        }
+        this.isRunning = true;
+        await this.redisClient.connect();
+        this.logger.info('TrialClient Redis connected');
+    }
+
+    async stop() {
+        try {
+            this.redisClient.disconnect();
+            this.logger.info('TrialClient Redis disconnected');
+        } finally {
+            this.isRunning = false;
+        }
+    }
+
+    private createRedisClient() {
+        return new Redis(this.REDIS_URL, {
+            lazyConnect: true,
+            disconnectTimeout: 10,
+        });
+    }
+
+    isTrialToken(token: Record<string, any>) {
+        return !!token.serviceRestrictions;
+    }
+
+    async assertTrial(token: Record<string, any>) {
+        const trial = await this.read(token.clientId);
+        if (!trial) {
+            throw new AccessForbidden('Trial access for token not configured');
+        }
+        return trial;
+    }
+
+    async requireServiceRestriction(trial: Trial, token: Record<string, any>, serviceName: string) {
+        const serviceRestriction = trial.serviceRestrictions.find((service: TrialServiceRestriction) => service.serviceName === serviceName);
+        const tokenServiceRestriction = token.serviceRestrictions.find((service: TokenServiceRestriction) => service.serviceName === serviceName);
+
+        if (!serviceRestriction || !tokenServiceRestriction) {
+            throw new AccessForbidden('Service access for token not configured');
+        }
+        if (serviceRestriction.requestCount > tokenServiceRestriction.requestLimit) {
+            throw new AccessForbidden('Trial token has exceeded request limit for service');
+        }
+        return serviceRestriction;
+    }
+
+    async incrementRequests(token: Record<string, any>, serviceName: string) {
+        const trial = await this.assertTrial(token);
+        const serviceRestriction = await this.requireServiceRestriction(trial, token, serviceName);
+        serviceRestriction.requestCount++;
```

Comment:
is this in memory only (per app instance)?

Reply:
no it'll be in the deployed redis cache, so it's keyed to `cache:framework:trialClient:<client_id>`, accessible across apps